### DISCUSSION TITLE
Accessing modules and functions in a separate file?

### DISCUSSION THREAD

**Discussion by hyanwong** [jupyterlite/jupyterlite#1072] [Q&A] [answered]
Also posted in SO [here](https://stackoverflow.com/questions/76162095), but moving to this forum in case it is bug to be flagged. I have a JupyterLite/pyodide session running a nodebook called `Workshop1.ipynb`. In the same (root) directory as that notebook, I have a file containing useful functions which also does some imports, e.g.

```python
# functions.py
import numpy as np

def a_useful_function(val):
    return np.array(val) * 2
```

I want to call `import functions` as the first line of my notebook, to use the functions defined in the `functions.py` file, but I then get

```
ModuleNotFoundError: The module 'numpy' is included in the Pyodide distribution, but it is not installed.
You can install it by calling:
  await micropip.install("numpy") in Python, or
  await pyodide.loadPackage("numpy") in JavaScript
```

Is it possible to load modules inside a file which is then imported into the notebook, without calling e.g. `import numpy` in the main notebook itself?

**Comment by jtpio** [comment, marked as answer]
Thanks @hyanwong for reporting.

It's very likely related to the way Pyodide loads the packages. If I recall correctly Pyodide looks for `import` statements when the code is executed and loads the appropriate packages. It's possible that it does not look at imports in modules.

Also for reference I just tried with the Xeus Python kernel and it seems to be working fine: https://github.com/jupyterlite/xeus-python-demo

![image](https://user-images.githubusercontent.com/591645/236217187-12e9c38f-3600-4897-917d-1de00af57783.png)

**Reply by rth** [reply]
I added a bit more context in https://stackoverflow.com/a/76208774/1791279. But yes, this is the expected behavior for Pyodide, you need to explicitly install requirements (or indirectly load them via first level imports). Xeus Python kernel works because numpy is being loaded by default, but this wouldn't work for an arbitrary package from what I understand.

**Reply by hyanwong** [reply]
Great, thanks for the answer.

**Reply by jtpio** [reply]
> Xeus Python kernel works because numpy is being loaded by default, but this wouldn't work for an arbitrary package from what I understand.

Would need to double check, but Xeus Python can also be used without numpy so I would assume it should work with other packages as long as they are included in the environment (via `empack`)?

**Reply by rth** [reply]
Yeah, I mean one way or another one needs to specify the packages that should to be installed in a given env. From what I understand in Xeus Python they would be specified when creating the env via empack, while in Pyodide you would need to specify them at runtime (provided someone built them). But anyway, an import somewhere in the filesystem will not install arbitrary packages, as requested here.
